fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java

```diff
@@ -638,30 +633,41 @@ boolean isErased() {
          */
         @Nullable
         @Override
-        public DescriptorProto buildDescriptor(@Nonnull final String typeName) {
+        public DescriptorProto buildDescriptor(final DynamicSchema.Builder dynamicSchemaBuilder, @Nonnull final String typeName) {
             Objects.requireNonNull(fields);
             final DescriptorProto.Builder recordMsgBuilder = DescriptorProto.newBuilder();
             recordMsgBuilder.setName(typeName);
 
             for (final Field field : fields) {
                 final Type fieldType = field.getFieldType();
                 final String fieldName = field.getFieldName();
-                fieldType.addProtoField(recordMsgBuilder,
-                        field.getFieldIndex(),
-                        fieldName,
-                        typeName(fieldName),
-                        FieldDescriptorProto.Label.LABEL_OPTIONAL);
+                if (!fieldType.isPrimitive()) {
+                    fieldType.addProtoField(dynamicSchemaBuilder, recordMsgBuilder,
+                            field.getFieldIndex(),
+                            fieldName,
+                            dynamicSchemaBuilder.addTypeAndGetName(fieldType),
+                            FieldDescriptorProto.Label.LABEL_OPTIONAL);
+                } else {
+                    fieldType.addProtoField(dynamicSchemaBuilder, recordMsgBuilder,
+                            field.getFieldIndex(),
+                            fieldName,
+                            "ignored", // gets the built-in PB primitive type name.
```

Comment:
I might've misunderstood what you meant, but the parameter `typeName` is not used in case of primitive types:

```java

@Override
public void addProtoField(@Nonnull final TypeRepository.Builder typeRepositoryBuilder,
                          @Nonnull final DescriptorProto.Builder descriptorBuilder,
                          final int fieldIndex,
                          @Nonnull final String fieldName,
                          @Nonnull final String ignored, // <--
                          @Nonnull final FieldDescriptorProto.Label label) {
    final FieldDescriptorProto.Type protoType = Objects.requireNonNull(getTypeCode().getProtoType());
    descriptorBuilder.addField(FieldDescriptorProto.newBuilder()
            .setNumber(fieldIndex)
            .setName(fieldName)
            .setType(protoType)
            .setLabel(label)
            .build());
}

```